File: widget/widget.js
/**
 * ContextOne Chat Widget
 * Embeddable AI chatbot widget for websites
 * Version: 1.0.0
 */
(function() {
  'use strict';

  // Configuration
  const config = window.ContextOneConfig || {};
  const projectId = config.projectId;
  const apiKey = config.apiKey;
  const apiUrl = config.apiUrl || 'http://localhost:8000';
  const primaryColor = config.primaryColor || '#3B82F6';
  const position = config.position || 'bottom-right';

  if (!projectId || !apiKey) {
    console.error('ContextOne: projectId and apiKey are required');
    return;
  }

  // State
  let isOpen = false;
  let conversationId = null;
  let messages = [];

  // Create widget HTML
  const createWidget = () => {
    const container = document.createElement('div');
    container.id = 'contextone-widget';
    container.className = `contextone-widget-${position}`;

    container.innerHTML = `
      <style>
        .contextone-widget-bottom-right {
          position: fixed;
          bottom: 20px;
          right: 20px;
          z-index: 9999;
          font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, sans-serif;
        }
        .contextone-widget-bottom-left {
          position: fixed;
          bottom: 20px;
          left: 20px;
          z-index: 9999;
          font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, sans-serif;
        }
        .contextone-bubble {
          width: 60px;
          height: 60px;
          border-radius: 50%;
          background: ${primaryColor};
          color: white;
          display: flex;
          align-items: center;
          justify-content: center;
          cursor: pointer;
          box-shadow: 0 4px 12px rgba(0,0,0,0.15);
          transition: transform 0.2s, box-shadow 0.2s;
        }
        .contextone-bubble:hover {
          transform: scale(1.05);
          box-shadow: 0 6px 16px rgba(0,0,0,0.2);
        }
        .contextone-bubble svg {
          width: 28px;
          height: 28px;
        }
        .contextone-window {
          position: fixed;
          bottom: 100px;
          width: 380px;
          height: 550px;
          background: white;
          border-radius: 12px;
          box-shadow: 0 8px 32px rgba(0,0,0,0.12);
          display: none;
          flex-direction: column;
          overflow: hidden;
        }
        .contextone-widget-bottom-right .contextone-window {
          right: 20px;
        }
        .contextone-widget-bottom-left .contextone-window {
          left: 20px;
        }
        .contextone-window.open {
          display: flex;
        }
        .contextone-header {
          background: ${primaryColor};
          color: white;
          padding: 16px 20px;
          display: flex;
          justify-content: space-between;
          align-items: center;
        }
        .contextone-header h3 {
          margin: 0;
          font-size: 16px;
          font-weight: 600;
        }
        .contextone-close {
          background: none;
          border: none;
          color: white;
          font-size: 24px;
          cursor: pointer;
          padding: 0;
          width: 30px;
          height: 30px;
          display: flex;
          align-items: center;
          justify-content: center;
          border-radius: 4px;
          transition: background 0.2s;
        }
        .contextone-close:hover {
          background: rgba(255,255,255,0.2);
        }
        .contextone-messages {
          flex: 1;
          overflow-y: auto;
          padding: 20px;
          background: #f9fafb;
        }
        .contextone-message {
          margin-bottom: 16px;
          display: flex;
          flex-direction: column;
        }
        .contextone-message.user {
          align-items: flex-end;
        }
        .contextone-message.assistant {
          align-items: flex-start;
        }
        .contextone-message-content {
          max-width: 80%;
          padding: 12px 16px;
          border-radius: 12px;
          font-size: 14px;
          line-height: 1.5;
        }
        .contextone-message.user .contextone-message-content {
          background: ${primaryColor};
          color: white;
        }
        .contextone-message.assistant .contextone-message-content {
          background: white;
          color: #1f2937;
          box-shadow: 0 1px 3px rgba(0,0,0,0.1);
        }
        .contextone-typing {
          display: none;
          padding: 12px 16px;
          background: white;
          border-radius: 12px;
          width: fit-content;
          box-shadow: 0 1px 3px rgba(0,0,0,0.1);
        }
        .contextone-typing.show {
          display: block;
        }
        .contextone-typing span {
          display: inline-block;
          width: 8px;
          height: 8px;
          border-radius: 50%;
          background: #9ca3af;
          margin: 0 2px;
          animation: contextone-bounce 1.4s infinite ease-in-out;
        }
        .contextone-typing span:nth-child(1) {
          animation-delay: -0.32s;
        }
        .contextone-typing span:nth-child(2) {
          animation-delay: -0.16s;
        }
        @keyframes contextone-bounce {
          0%, 80%, 100% {
            transform: scale(0);
          }
          40% {
            transform: scale(1);
          }
        }
        .contextone-input-container {
          padding: 16px;
          background: white;
          border-top: 1px solid #e5e7eb;
          display: flex;
          gap: 8px;
        }
        .contextone-input {
          flex: 1;
          border: 1px solid #e5e7eb;
          border-radius: 8px;
          padding: 10px 12px;
          font-size: 14px;
          outline: none;
          transition: border-color 0.2s;
        }
        .contextone-input:focus {
          border-color: ${primaryColor};
        }
        .contextone-send {
          background: ${primaryColor};
          color: white;
          border: none;
          border-radius: 8px;
          padding: 10px 20px;
          font-size: 14px;
          font-weight: 600;
          cursor: pointer;
          transition: opacity 0.2s;
        }
        .contextone-send:hover:not(:disabled) {
          opacity: 0.9;
        }
        .contextone-send:disabled {
          opacity: 0.5;
          cursor: not-allowed;
        }
        .contextone-sources {
          font-size: 12px;
          color: #6b7280;
          margin-top: 8px;
        }
        .contextone-sources a {
          color: ${primaryColor};
          text-decoration: none;
        }
      </style>

      <div class="contextone-bubble" id="contextone-bubble">
        <svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke="currentColor">
          <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M8 12h.01M12 12h.01M16 12h.01M21 12c0 4.418-4.03 8-9 8a9.863 9.863 0 01-4.255-.949L3 20l1.395-3.72C3.512 15.042 3 13.574 3 12c0-4.418 4.03-8 9-8s9 3.582 9 8z" />
        </svg>
      </div>

      <div class="contextone-window" id="contextone-window">
        <div class="contextone-header">
          <h3>Chat with us</h3>
          <button class="contextone-close" id="contextone-close">&times;</button>
        </div>
        <div class="contextone-messages" id="contextone-messages">
          <div class="contextone-message assistant">
            <div class="contextone-message-content">
              Hi! How can I help you today?
            </div>
          </div>
        </div>
        <div class="contextone-typing" id="contextone-typing">
          <span></span>
          <span></span>
          <span></span>
        </div>
        <div class="contextone-input-container">
          <input
            type="text"
            class="contextone-input"
            id="contextone-input"
            placeholder="Ask a question..."
            autocomplete="off"
          />
          <button class="contextone-send" id="contextone-send">Send</button>
        </div>
      </div>
    `;

    document.body.appendChild(container);
  };

  // Add message to UI
  const addMessage = (role, content, sources) => {
    const messagesContainer = document.getElementById('contextone-messages');
    const messageDiv = document.createElement('div');
    messageDiv.className = `contextone-message ${role}`;

    let sourcesHtml = '';
    if (sources && sources.length > 0) {
      sourcesHtml = `<div class="contextone-sources">Sources: ${sources.map(s => s.doc_name).join(', ')}</div>`;
    }

    messageDiv.innerHTML = `
      <div class="contextone-message-content">${content}</div>
      ${sourcesHtml}
    `;

    messagesContainer.appendChild(messageDiv);
    messagesContainer.scrollTop = messagesContainer.scrollHeight;
  };

  // Show typing indicator
  const showTyping = (show) => {
    const typing = document.getElementById('contextone-typing');
    if (show) {
      typing.classList.add('show');
    } else {
      typing.classList.remove('show');
    }
  };

  // Send message to API
  const sendMessage = async (message) => {
    const sendBtn = document.getElementById('contextone-send');
    const input = document.getElementById('contextone-input');

    // Disable input
    sendBtn.disabled = true;
    input.disabled = true;

    // Add user message to UI
    addMessage('user', message);

    // Clear input
    input.value = '';

    // Show typing
    showTyping(true);

    try {
      const response = await fetch(`${apiUrl}/api/widget/${projectId}/chat`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-API-Key': apiKey
        },
        body: JSON.stringify({
          query: message,
          conversation_id: conversationId
        })
      });

      if (!response.ok) {
        throw new Error('Failed to get response');
      }

      const data = await response.json();

      // Update conversation ID
      conversationId = data.conversation_id;

      // Hide typing
      showTyping(false);

      // Add assistant response
      addMessage('assistant', data.answer, data.sources);

    } catch (error) {
      console.error('ContextOne Error:', error);
      showTyping(false);
      addMessage('assistant', 'Sorry, I encountered an error. Please try again.');
    } finally {
      // Re-enable input
      sendBtn.disabled = false;
      input.disabled = false;
      input.focus();
    }
  };

  // Event handlers
  const attachEvents = () => {
    const bubble = document.getElementById('contextone-bubble');
    const window = document.getElementById('contextone-window');
    const closeBtn = document.getElementById('contextone-close');
    const sendBtn = document.getElementById('contextone-send');
    const input = document.getElementById('contextone-input');

    bubble.addEventListener('click', () => {
      isOpen = !isOpen;
      window.classList.toggle('open', isOpen);
      if (isOpen) {
        input.focus();
      }
    });

    closeBtn.addEventListener('click', () => {
      isOpen = false;
      window.classList.remove('open');
    });

    sendBtn.addEventListener('click', () => {
      const message = input.value.trim();
      if (message) {
        sendMessage(message);
      }
    });

    input.addEventListener('keypress', (e) => {
      if (e.key === 'Enter') {
        const message = input.value.trim();
        if (message) {
          sendMessage(message);
        }
      }
    });
  };

  // Initialize widget
  const init = () => {
    if (document.readyState === 'loading') {
      document.addEventListener('DOMContentLoaded', () => {
        createWidget();
        attachEvents();
      });
    } else {
      createWidget();
      attachEvents();
    }
  };

  init();
})();

File: backend/src/routers/widget.py
from fastapi import APIRouter, HTTPException, Depends, Header
from fastapi.responses import StreamingResponse
from supabase import Client
from ..dependencies.database import get_supabase

router = APIRouter()

@router.get("/widget/{project_id}/config")
async def get_widget_config(project_id: str, x_api_key: str = Header(None), supabase: Client = Depends(get_supabase)):
    if not x_api_key:
        raise HTTPException(status_code=401, detail="API key required")
    
    if not supabase:
        raise HTTPException(status_code=500, detail="Database not configured")
    
    key_response = supabase.table("api_keys").select("tenant_id").eq("key", x_api_key).eq("is_active", True).execute()
    if not key_response.data:
        raise HTTPException(status_code=401, detail="Invalid API key")
    
    tenant_id = key_response.data[0]["tenant_id"]
    
    project_response = supabase.table("projects").select("*").eq("project_id", project_id).eq("tenant_id", tenant_id).execute()
    if not project_response.data:
        raise HTTPException(status_code=404, detail="Project not found")
    
    project = project_response.data[0]
    
    return {
        "project_id": project["project_id"],
        "name": project["name"],
        "welcome_message": project["welcome_message"],
        "ui_config": project["ui_config"]
    }

@router.get("/widget.js")
async def get_widget_js():
    widget_js = '''
(function() {
    const WIDGET_STYLES = `
        .ctx-widget-container { position: fixed; z-index: 9999; font-family: 'Inter', sans-serif; }
        .ctx-widget-container.bottom-right { bottom: 20px; right: 20px; }
        .ctx-widget-container.bottom-left { bottom: 20px; left: 20px; }
        .ctx-widget-button { width: 60px; height: 60px; border-radius: 50%; border: none; cursor: pointer; display: flex; align-items: center; justify-content: center; box-shadow: 0 4px 12px rgba(0,0,0,0.3); transition: transform 0.2s; }
        .ctx-widget-button:hover { transform: scale(1.05); }
        .ctx-widget-button svg { width: 28px; height: 28px; fill: white; }
        .ctx-widget-chat { position: absolute; bottom: 70px; right: 0; width: 380px; height: 500px; background: #0B0E14; border: 1px solid #2D3748; border-radius: 12px; display: none; flex-direction: column; overflow: hidden; box-shadow: 0 10px 40px rgba(0,0,0,0.5); }
        .ctx-widget-chat.open { display: flex; }
        .ctx-widget-header { padding: 16px; background: #151921; border-bottom: 1px solid #2D3748; display: flex; align-items: center; justify-content: space-between; }
        .ctx-widget-header h3 { margin: 0; color: white; font-size: 16px; font-weight: 600; }
        .ctx-widget-close { background: none; border: none; color: #9CA3AF; cursor: pointer; font-size: 20px; }
        .ctx-widget-messages { flex: 1; padding: 16px; overflow-y: auto; display: flex; flex-direction: column; gap: 12px; }
        .ctx-widget-message { max-width: 85%; padding: 12px 16px; border-radius: 12px; font-size: 14px; line-height: 1.5; }
        .ctx-widget-message.user { background: #6366F1; color: white; align-self: flex-end; border-bottom-right-radius: 4px; }
        .ctx-widget-message.assistant { background: #1F2937; color: #E5E7EB; align-self: flex-start; border-bottom-left-radius: 4px; }
        .ctx-widget-input-area { padding: 12px; border-top: 1px solid #2D3748; display: flex; gap: 8px; }
        .ctx-widget-input { flex: 1; padding: 10px 14px; border: 1px solid #2D3748; border-radius: 8px; background: #151921; color: white; font-size: 14px; outline: none; }
        .ctx-widget-input:focus { border-color: #6366F1; }
        .ctx-widget-send { padding: 10px 16px; background: #6366F1; color: white; border: none; border-radius: 8px; cursor: pointer; font-weight: 500; }
        .ctx-widget-send:hover { background: #5558E3; }
        .ctx-widget-send:disabled { opacity: 0.5; cursor: not-allowed; }
    `;

    class ContextOneWidget {
        constructor(config) {
            this.projectId = config.projectId;
            this.apiKey = config.apiKey;
            this.apiUrl = config.apiUrl || window.location.origin;
            this.position = config.position || 'bottom-right';
            this.primaryColor = config.primaryColor || '#6366F1';
            this.messages = [];
            this.init();
        }

        init() {
            this.injectStyles();
            this.createWidget();
            this.loadConfig();
        }

        injectStyles() {
            const style = document.createElement('style');
            style.textContent = WIDGET_STYLES;
            document.head.appendChild(style);
        }

        createWidget() {
            this.container = document.createElement('div');
            this.container.className = `ctx-widget-container ${this.position}`;
            this.container.innerHTML = `
                <div class="ctx-widget-chat">
                    <div class="ctx-widget-header">
                        <h3>AI Assistant</h3>
                        <button class="ctx-widget-close">&times;</button>
                    </div>
                    <div class="ctx-widget-messages"></div>
                    <div class="ctx-widget-input-area">
                        <input type="text" class="ctx-widget-input" placeholder="Type your message..." />
                        <button class="ctx-widget-send">Send</button>
                    </div>
                </div>
                <button class="ctx-widget-button" style="background: ${this.primaryColor}">
                    <svg viewBox="0 0 24 24"><path d="M20 2H4c-1.1 0-2 .9-2 2v18l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm0 14H6l-2 2V4h16v12z"/></svg>
                </button>
            `;
            document.body.appendChild(this.container);
            this.bindEvents();
        }

        bindEvents() {
            const btn = this.container.querySelector('.ctx-widget-button');
            const chat = this.container.querySelector('.ctx-widget-chat');
            const close = this.container.querySelector('.ctx-widget-close');
            const input = this.container.querySelector('.ctx-widget-input');
            const send = this.container.querySelector('.ctx-widget-send');

            btn.onclick = () => chat.classList.toggle('open');
            close.onclick = () => chat.classList.remove('open');
            send.onclick = () => this.sendMessage();
            input.onkeypress = (e) => e.key === 'Enter' && this.sendMessage();
        }

        async loadConfig() {
            try {
                const res = await fetch(`${this.apiUrl}/api/widget/${this.projectId}/config`, {
                    headers: { 'X-API-Key': this.apiKey }
                });
                if (res.ok) {
                    const config = await res.json();
                    this.container.querySelector('.ctx-widget-header h3').textContent = config.name || 'AI Assistant';
                    if (config.welcome_message) {
                        this.addMessage('assistant', config.welcome_message);
                    }
                }
            } catch (e) { console.error('Widget config error:', e); }
        }

        addMessage(role, content) {
            const messages = this.container.querySelector('.ctx-widget-messages');
            const msg = document.createElement('div');
            msg.className = `ctx-widget-message ${role}`;
            msg.textContent = content;
            messages.appendChild(msg);
            messages.scrollTop = messages.scrollHeight;
        }

        async sendMessage() {
            const input = this.container.querySelector('.ctx-widget-input');
            const send = this.container.querySelector('.ctx-widget-send');
            const query = input.value.trim();
            if (!query) return;

            this.addMessage('user', query);
            input.value = '';
            send.disabled = true;

            try {
                const res = await fetch(`${this.apiUrl}/api/widget/${this.projectId}/chat`, {
                    method: 'POST',
                    headers: { 'Content-Type': 'application/json', 'X-API-Key': this.apiKey },
                    body: JSON.stringify({ query, conversation_history: [] })
                });
                const data = await res.json();
                this.addMessage('assistant', data.answer || 'Sorry, I could not process that.');
            } catch (e) {
                this.addMessage('assistant', 'Error connecting to the server.');
            }
            send.disabled = false;
        }
    }

    window.ContextOneWidget = ContextOneWidget;
})();
'''
    return StreamingResponse(
        iter([widget_js]),
        media_type="application/javascript",
        headers={"Content-Disposition": "inline; filename=widget.js"}
    )